drivers/adc/adc_gecko.c

```diff
@@ -0,0 +1,376 @@
+{
+	const struct adc_gecko_cfg *config = DEV_CFG(dev);
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	LOG_DBG("Starting conversion");
+
+	/* Start single conversion */
+	ADC_Start(adc, adcStartSingle);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_gecko_data *data =
+		CONTAINER_OF(ctx, struct adc_gecko_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	adc_gecko_start_conversion(data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_gecko_data *data =
+		CONTAINER_OF(ctx, struct adc_gecko_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void adc_gecko_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct adc_gecko_cfg *config = DEV_CFG(dev);
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	ADC_TypeDef *adc = config->base;
+
+	*data->buffer++ = (u16_t)ADC_DataSingleGet(adc);
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	LOG_DBG("ISR triggered.");
+}
+
+static int adc_gecko_read(struct device *dev,
+			  const struct adc_sequence *sequence)
+{
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_gecko_read_async(struct device *dev,
+				const struct adc_sequence *sequence,
+				struct k_poll_signal *async)
+{
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif
+
+static int adc_gecko_check_acq_time(u16_t acq_time)
+{
+	for (int i = 0; i < 10; i++) {
+		if (acq_time == ADC_ACQ_TIME(ADC_ACQ_TIME_TICKS,
+					     acq_time_single_tbl[i])) {
+			return i;
+		}
+	}
+
+	if (acq_time == ADC_ACQ_TIME_DEFAULT) {
+		return 0;
+	}
+
+	LOG_ERR("Conversion time not supported.");
+	return -EINVAL;
+}
+
+
+static int adc_gecko_channel_setup(struct device *dev,
+				   const struct adc_channel_cfg *channel_cfg)
+{
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	struct adc_gecko_ch_cfg *config_ch = NULL;
+	int acq_time_index;
+
+	if (channel_cfg->channel_id >= GECKO_CHANNEL_COUNT) {
+		LOG_ERR("Channel %d is not valid", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not (yet) supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	/* Setup input */
+	config_ch = CH_CFG(channel_cfg->channel_id);
+	config_ch->input_positive = channel_cfg->input_positive;
+
+	/* Setup acquistion time */
+	acq_time_index = adc_gecko_check_acq_time(
+		channel_cfg->acquisition_time);
+	if (acq_time_index < 0) {
+		return acq_time_index;
```

Comment:
This is a bit confusing. You're actually returning an error here, not an acquisition time index.

Reply:
Agreed, `return acq_time_index;` is now replaced by a direct `return -EINVAL;` as everywhere else in the code.